src/librustc_resolve/lib.rs

```diff
@@ -2605,6 +2605,92 @@ impl<'a> Resolver<'a> {
                     mk_struct_span_error(false)
                 };
 
+                err.span_label(ident.span, &format!("this {} is private", binding.res().descr()));
+
+                // Display the chain of re-exports through to the original def for cases where the
+                // `use` is private but the def is public.
+                let mut imported = false;
+                let mut binding = binding;
+                loop {
+                    match binding.kind {
+                        NameBindingKind::Res(res, _is_macro_export) => {
+                            match (res, imported, binding.vis) {
+                                (Res::Def(_, def_id), true, ty::Visibility::Public) => {
+                                    // FIXME: we should verify that this def is actually
+                                    // reachable from the user's crate, as the parent modules
+                                    // of this ADT might be private.
```

Comment:
@petrochenkov would you have some advice on how to properly check the visibility of a given ADT here? I'm guessing I would have to ferry some scope information in `PrivacyError`.

Reply:
This issue doesn't apply to the simpler version of this PR (https://github.com/rust-lang/rust/pull/67951#issuecomment-573343222) which I'd like to see implemented first.
I'm not sure we'll need anything beyond that.